File: fubon.wms/src/webapp/assets/txn/CMMGR015/CMMGR015.js
/*
  交易畫面邏輯定義
  請修改 Controller 名稱，以符合交易畫面的定義
 */
'use strict';
eSoafApp.controller('CMMGR015_Controller', function($rootScope, $scope, $controller, $confirm, socketService, ngDialog, projInfoService, getParameter) {
	$controller('BaseController', {
		$scope : $scope
	});
	$scope.controllerName = "CMMGR015_Controller";

	// init
	$scope.init = function() {
		$scope.inputVO.code = '';
		$scope.inputVO.text = '';
		$scope.inputVO.sqlStr = '';
		$scope.inputVO.sqlResult = '';
	}
	$scope.init();

	// 查詢
	$scope.inquire = function() {
		$scope.sendRecv('CMMGR015', 'inquire', 'com.systex.jbranch.app.server.fps.cmmgr015.CMMGR015InputVO', $scope.inputVO, function(tota, isError) {
			if (!isError) {
				$scope.outputVO = tota[0].body;
				$scope.i18nlist = tota[0].body.resultList;
				$scope.init();
			} else {
				$scope.showMsg('查詢失敗');
			}
		});

	}

	// 查詢
	$scope.inquireSQL = function() {
		$scope.sendRecv('CMMGR015', 'inquireSQL', 'com.systex.jbranch.app.server.fps.cmmgr015.CMMGR015InputVO', $scope.inputVO, function(tota, isError) {
			if (!isError) {
				$scope.outputVO = tota[0].body;
				$scope.inputVO.sqlResult = tota[0].body.sqlResult;
			} else {
				$scope.showMsg('查詢失敗');
			}
		});

	}
	// 新增
	$scope.insert = function() {
		
		$confirm({
			title : '新增',
			text : '是否新增訊息代碼: ' + $scope.inputVO.code + '訊息內容: ' + $scope.inputVO.text +'?'
		}, {
			size : 'sm'
		}).then(function() {
			
			$scope.sendRecv('CMMGR015', 'insert', 'com.systex.jbranch.app.server.fps.cmmgr015.CMMGR015InputVO', $scope.inputVO, function(tota, isError) {
				if (!isError) {
					$scope.showMsg('新增成功');
					$scope.inquire();
				} else {
					$scope.showMsg('新增失敗');
				}
			});
			
		});
		
	}
	
	// 修改
	$scope.update = function() {
		
		$confirm({
			title : '修改',
			text : '是否將訊息代碼: ' + $scope.choosecode + '訊息內容: ' + $scope.choosetext + 
			'修改成訊息代碼: ' + $scope.inputVO.code + '訊息內容: ' + $scope.inputVO.text +'?'
		}, {
			size : 'sm'
		}).then(function() {
			
			$scope.sendRecv('CMMGR015', 'update', 'com.systex.jbranch.app.server.fps.cmmgr015.CMMGR015InputVO', $scope.inputVO, function(tota, isError) {
				if (!isError) {
					$scope.showMsg('修改成功');
					$scope.inquire();
				} else {
					$scope.showMsg('修改失敗');
				}
			});
			
			
		});
		
	}
	
	// 修改
	$scope.updateSQL = function() {
		$scope.sendRecv('CMMGR015', 'updateSQL', 'com.systex.jbranch.app.server.fps.cmmgr015.CMMGR015InputVO', $scope.inputVO, function(tota, isError) {
			if (!isError) {
				$scope.inputVO.sqlResult = tota[0].body.sqlResult;
			} else {
				$scope.inputVO.sqlResult = tota[0].body.sqlResult;
			}
		});
	}
	// 刪除
	
	
	$scope.Delete = function() {
		$confirm({
			title : '刪除',
			text : '是否刪除訊息代碼: ' + $scope.inputVO.code + '訊息內容: ' + $scope.inputVO.text +'?'
		}, {
			size : 'sm'
		}).then(function() {
			$scope.sendRecv('CMMGR015', 'delete', 'com.systex.jbranch.app.server.fps.cmmgr015.CMMGR015InputVO', $scope.inputVO, function(tota, isError) {
				if (!isError) {
					$scope.showMsg('刪除成功');
					$scope.inquire();
				} else {
					$scope.showMsg('刪除失敗');
				}
			});
		});
	
	}
	
	
	//選擇訊息代號或訊息內容，會自動帶入上方欄位供使用者修改或刪除
	$scope.choose = function(row){
		$scope.choosecode = row.CODE;
		$scope.choosetext = row.CODE;
		
		$scope.inputVO.code = row.CODE;
		$scope.inputVO.text = row.TEXT;
	}

});

File: fubon.wms/src/main/java/com/systex/jbranch/app/server/fps/cmmgr015/CMMGR015.java
package com.systex.jbranch.app.server.fps.cmmgr015;

import com.systex.jbranch.fubon.commons.FubonWmsBizLogic;
import com.systex.jbranch.platform.common.dataaccess.delegate.DataAccessManager;
import com.systex.jbranch.platform.common.dataaccess.query.QueryConditionIF;
import com.systex.jbranch.platform.common.errHandle.APException;
import com.systex.jbranch.platform.common.errHandle.DAOException;
import com.systex.jbranch.platform.common.errHandle.JBranchException;
import com.systex.jbranch.platform.common.platformdao.table.TBSYSI18NPK;
import com.systex.jbranch.platform.common.platformdao.table.TBSYSI18NVO;
import com.systex.jbranch.platform.util.IPrimitiveMap;
import org.apache.commons.lang.StringUtils;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import java.math.BigDecimal;
import java.util.Arrays;
import java.util.List;
import java.util.Map;

/**
 * 訊息維護
 *
 * @author SamTu
 * @date 20180719
 *
 */
@Component("cmmgr015")
@Scope("request")
public class CMMGR015 extends FubonWmsBizLogic {
	private DataAccessManager dam;
	private QueryConditionIF condition;

	/** 取得 {@link DataAccessManager} 物件 */
	private void getDam() {
		dam = this.getDataAccessManager();
	}

	/** 取得 {@link QueryConditionIF} 物件 */
	private void getCondition() throws DAOException, JBranchException {
		condition = dam.getQueryCondition(DataAccessManager.QUERY_LANGUAGE_TYPE_VAR_SQL);
	}

	// 查詢
	public void inquire(Object body, IPrimitiveMap header) throws DAOException, JBranchException {

		CMMGR015InputVO inputVO = (CMMGR015InputVO) body;

		// System.out.println(cmmgr015InputVO.getCode()+","+cmmgr015InputVO.getText());
		getDam();
		getCondition();

		condition.setQueryString(new StringBuffer().append("select CODE, TEXT ").append("from TBSYSI18N ").toString());
		dam.exeQuery(condition);
		CMMGR015OutputVO returnVO = new CMMGR015OutputVO();
		returnVO.setResultList(dam.exeQuery(condition));
		sendRtnObject(returnVO);
	}

	// 新增
	public void insert(Object body, IPrimitiveMap header) throws DAOException, JBranchException {
		CMMGR015InputVO inputVO = (CMMGR015InputVO) body;

		// System.out.println(cmmgr015InputVO.getCode()+","+cmmgr015InputVO.getText());
		getDam();
		getCondition();

		TBSYSI18NVO vo = new TBSYSI18NVO();
		TBSYSI18NPK pk = new TBSYSI18NPK();
		pk.setLOCALE("zh-tw");
		pk.setCODE(inputVO.getCode());
		TBSYSI18NVO queryVO = new TBSYSI18NVO();// 用于判斷是否已經存在相同主鍵
		// 先查找是否已經存在同樣主鍵的數據
		queryVO = (TBSYSI18NVO) dam.findByPKey(TBSYSI18NVO.TABLE_UID, pk);
		if (queryVO != null) {
			throw new APException("ehl_01_common_005");
		} else {
			vo.setcomp_id(pk);
			vo.setTEXT(inputVO.getText());
			vo.setTYPE("T");
			dam.create(vo);
		}
		this.sendRtnObject(null);
	}

	// 修改
	public void update(Object body, IPrimitiveMap header) throws DAOException, JBranchException {
		CMMGR015InputVO inputVO = (CMMGR015InputVO) body;

		// System.out.println(cmmgr015InputVO.getCode()+","+cmmgr015InputVO.getText());
		getDam();
		getCondition();

		TBSYSI18NVO vo = new TBSYSI18NVO();
		TBSYSI18NPK pk = new TBSYSI18NPK();
		pk.setLOCALE("zh-tw");
		pk.setCODE(inputVO.getCode());

		vo = (TBSYSI18NVO) dam.findByPKey(
				TBSYSI18NVO.TABLE_UID, pk);
		if (vo != null) {
			vo.setTEXT(inputVO.getText());
			vo.setTYPE("T");
			dam.update(vo);
		}
		this.sendRtnObject(null);


	}

	// 刪除
	public void delete(Object body, IPrimitiveMap header) throws DAOException, JBranchException {
		CMMGR015InputVO inputVO = (CMMGR015InputVO) body;

		// System.out.println(cmmgr015InputVO.getCode()+","+cmmgr015InputVO.getText());
		getDam();
		getCondition();

		TBSYSI18NVO vo = new TBSYSI18NVO();
		TBSYSI18NPK pk = new TBSYSI18NPK();
		pk.setLOCALE("zh-tw");
		pk.setCODE(inputVO.getCode());

		vo = (TBSYSI18NVO) dam.findByPKey(TBSYSI18NVO.TABLE_UID, pk);
		if (vo != null) {
			dam.delete(vo);
		} else {
			// 顯示資料不存在
			throw new APException("ehl_01_common_001");
		}
		this.sendRtnObject(null);
	}

	  public void inquireSQL(Object body, IPrimitiveMap header)
			    throws DAOException, JBranchException
			  {
			    CMMGR015InputVO inputVO = ((CMMGR015InputVO)body);

			    getDam();
			    getCondition();

			  	String sqlStr = inputVO.getSqlStr();
			  	validFetchSize(sqlStr);
			  	this.logger.info(sqlStr);
			    this.condition.setQueryString(sqlStr);

			    Map<String, Object> resultMap = null;
			    StringBuffer sb = new StringBuffer("");
			    Object[] keys = null;
			    List<Map<String, Object>> resultList = this.dam.exeQuery(this.condition);
			    if ((resultList == null) || (resultList.size() == 0)) {
			      return;
			    }
			    resultMap = (Map)resultList.get(0);
			    keys = resultMap.keySet().toArray();
			    Arrays.sort(keys);

			    String colName = "";
			    for (int col = 0; col < keys.length; col++)
			    {
			      colName = keys[col].toString();
			      sb.append(colName).append("\t");
			    }
			    sb.append("\r\n");
			    for (int row = 0; row < resultList.size(); row++)
			    {
			      for (int col = 0; col < keys.length; col++)
			      {
			        colName = keys[col].toString();
			        sb.append(isNotNull(((Map)resultList.get(row)).get(colName))).append("\t");
			      }
			      sb.append("\r\n");
			    }

				CMMGR015OutputVO returnVO = new CMMGR015OutputVO();
			    returnVO.setSqlResult(sb.toString());
			    sendRtnObject(returnVO);
			  }

	private void validFetchSize(String sqlStr) throws JBranchException {
		// count 筆數不需要 valid
		if (StringUtils.containsIgnoreCase(sqlStr, "count("))
			return;

		BigDecimal fetchSizeLimit = getFetchSizeLimit();
		BigDecimal querySize = getQuerySize(sqlStr);
		if (fetchSizeLimit.compareTo(querySize) < 0) {
			throw new APException(String.format("查詢筆數共 %d 筆，已超過限制筆數 %d！請調整查詢語法，謝謝！",querySize.intValue(), fetchSizeLimit.intValue()));
		}
	}

	private BigDecimal getFetchSizeLimit() throws JBranchException {
		condition.setQueryString("select PARAM_CODE from TBSYSPARAMETER where PARAM_TYPE = 'CMMGR015.QRY' and PARAM_NAME = 'FETCH_SIZE' ");
		List<Map<String, String>> data= dam.exeQueryWithoutSort(condition);
		if (data.size() > 0) {
			String fetchsize = data.get(0).get("PARAM_CODE");
			if (fetchsize.matches("\\d+"))
				return new BigDecimal(fetchsize);
		}
		throw new APException("請確認 CMMGR015.QRY-FETCH_SIZE 參數是否設定正確！");
	}

	private BigDecimal getQuerySize(String sqlStr) throws JBranchException {
		String countSql = String.format("select count(1) CNT from (%s)", sqlStr);
		condition.setQueryString(countSql);
		List<Map<String, BigDecimal>> data = dam.exeQueryWithoutSort(condition);
		return data.get(0).get("CNT");
	}

	public void updateSQL(Object body, IPrimitiveMap header)
	    throws DAOException, JBranchException
	  {
	    CMMGR015InputVO inputVO = ((CMMGR015InputVO)body);

	    getDam();
	    getCondition();
	    this.logger.info(inputVO.getSqlStr());
	    String sqlStr = inputVO.getSqlStr();

		CMMGR015OutputVO outputVO = new CMMGR015OutputVO();
	    if (!sqlStr.startsWith("##"))
	    {
	      outputVO.setSqlResult("無法執行更新");
	    }
	    else
	    {
	      sqlStr = sqlStr.replaceFirst("##", "");
	      this.condition.setQueryString(sqlStr);
	      Integer cnt = Integer.valueOf(this.dam.exeUpdate(this.condition));
	      outputVO.setSqlResult("執行更新結果:" + cnt + "筆");
	    }
	    sendRtnObject(outputVO);
	  }

	  private String isNotNull(Object value)
			  {
			    try
			    {
			      return value.toString().trim();
			    }
			    catch (Exception localException) {}
			    return "";
			  }
}
